addons/stock/tests/test_quant_inventory_mode.py

```diff
@@ -294,3 +294,42 @@ def test_sn_warning(self):
         warning = dupe_sn._onchange_serial_number()
         self.assertTrue(warning, 'Reuse of existing serial number not detected')
         self.assertEqual(list(warning.keys())[0], 'warning', 'Warning message was not returned')
+
+    def test_reverse_inventory_adjustment(self):
+        """Try to reverse inventory adjustment"""
+        default_wh = self.env['stock.warehouse'].search([('company_id', '=', self.env.company.id)], limit=1)
+        default_stock_location = default_wh.lot_stock_id
+        quant = self.Quant.create({
+            'product_id': self.product.id,
+            'location_id': default_stock_location.id,
+            'inventory_quantity': 100,
+        })
+        quant.action_apply_inventory()
+        action = quant.action_inventory_history()
+        move_line = self.env['stock.move.line'].search(action['domain'])
+        self.assertEqual(self.product.qty_available, 100, "Before reverse inventory adjustment qty is 100")
+        move_line.action_revert_inventory()
+        self.assertEqual(self.product.qty_available, 0, "After reverse inventory adjustment qty is not zero")
+
+    def test_multi_reverse_inventory_adjustment(self):
+        """Try to reverse inventory adjustment with multiple lines"""
+
+        default_wh = self.env['stock.warehouse'].search([('company_id', '=', self.env.company.id)], limit=1)
+        default_stock_location = default_wh.lot_stock_id
+        quant_1 = self.Quant.create({
+            'product_id': self.product.id,
+            'location_id': default_stock_location.id,
+            'inventory_quantity': 100,
+        })
+        quant_1.action_apply_inventory()
+        quant_2 = self.Quant.create({
+            'product_id': self.product.id,
+            'location_id': default_stock_location.id,
+            'inventory_quantity': 150,
+        })
+        quant_2.action_apply_inventory()
+        action = quant_2.action_inventory_history()
+        move_line = self.env['stock.move.line'].search(action['domain'])
+        self.assertEqual(self.product.qty_available, 150, "Before reverse multi inventory adjustment qty is 150")
```

Comment:
You made and applied 2 quants, 100 + 150, so shouldn't qty_available = 250?

Reply:
it's count qty so first it's 100 qty and after 150 so it's added 50 qty only

Reply:
I think this looks weird to me because you should reuse quant_1 instead of having 2 quants in the same location for a non-tracked product (we normally do not allow this and/or they are auto-merged at different times). I.e.

```
        quant_1 = self.Quant.create({
            'product_id': self.product.id,
            'location_id': default_stock_location.id,
            'inventory_quantity': 100,
        })
        quant_1.action_apply_inventory()
        quant_1.inventory_quantity = 150
        quant_1.action_apply_inventory()
        action = quant_1.action_inventory_history()
        move_lines = self.env['stock.move.line'].search(action['domain'])
```

For example, doing action_apply_inventory() with 2 different quants like your test does within the Odoo shell results in the qty_available = 250 when I tested it. So it's probably safest if you reuse your original quant to avoid unexpected behavior.